VariantShredding.md

```diff
@@ -25,276 +25,302 @@
- optional binary variant_value;
- optional group object {
-  optional group a {
-   optional binary variant_value;
-   optional int64 typed_value;
-  }
-  optional group b {
-   optional binary variant_value;
-   optional group object {
-    optional group c {
-      optional binary variant_value;
-      optional binary typed_value (STRING);
-    }
-   }
-  }
- }
+required group measurement (VARIANT) {
+  required binary metadata;
+  optional binary value;
+  optional int64 typed_value;
 }
 ```
 
-| Variant Value | Top-level variant_value | b.variant_value | a.typed_value | a.variant_value | b.object.c.typed_value | b.object.c.variant_value | Notes | 
-|---------------|-------------------------|-----------------|---------------|-----------------|------------------------|--------------------------|-------|
-| {a: 123, b: {c: “hello”}} | null | null | 123 | null | hello | null | All values shredded |
-| {a: 1.23, b: {c: “123”}} | null | null | null | 1.23 | 123 | null | a is not an integer |
-| {a: 123, b: {c: null}} | null | null | null | 123 | null | null | b.object.c set to non-null to indicate VariantNull |
-| {a: 123, b: {} | null | null | null | 123 | null | null | b.object.c set to null, to indicate that c is missing |
-| {a: 123, d: 456} | {d: 456} | null | 123 | null | null | null | Extra field d is stored as variant_value |
-| [{a: 1, b: {c: 2}}, {a: 3, b: {c: 4}}] | [{a: 1, b: {c: 2}}, {a: 3, b: {c: 4}}] | null | null | null | null | null | Not an object |
+The series of measurements `34, null, "n/a", 100` would be stored as:
 
-# Parquet Layout
+| Value   | `metadata`       | `value`               | `typed_value` |
+|---------|------------------|-----------------------|---------------|
+| 34      | `01 00` v1/empty | null                  | `34`          |
+| null    | `01 00` v1/empty | `00` (null)           | null          |
+| "n/a"   | `01 00` v1/empty | `13 6E 2F 61` (`n/a`) | null          |
+| 100     | `01 00` v1/empty | null                  | `100`         |
 
-The `array` and `object` fields represent Variant array and object types, respectively.
-Arrays must use the three-level list structure described in https://github.com/apache/parquet-format/blob/master/LogicalTypes.md.
+Both `value` and `typed_value` are optional fields used together to encode a single value.
+Values in the two fields must be interpreted according to the following table:
 
-An `object` field must be a group.
-Each field name of this inner group corresponds to the Variant value's object field name.
-Each inner field's type is a recursively shredded variant value: that is, the fields of each object field must be one or more of `object`, `array`, `typed_value` or `variant_value`.
+| `value`  | `typed_value` | Meaning                                                  |
+|----------|---------------|----------------------------------------------------------|
+| null     | null          | The value is missing                                     |
```

Comment:
This is something I was a little confused about before. How do I differentiate between
```json
{ "foo" : { "x" : "null" }}
``` 
```json
{ "foo" : { }}
```

I may be missing something here but i'm trying to understand an empty vs missing shredding representation

Reply:
Ok so if we have no difference between

||  foo.typed_value.x.typed_value | foo.typed_value.x.value |
|--| --|--|
|foo : { x : "null" } | null                                                | null |
|foo : { y : "bar" } | null                                                | null |

Reply:
No, in the first case, `foo.typed_value.x.value` would be non-null, containing the variant value null (0x00). (Again, assuming you mean JSON null in your example, and not the string "null").

In the second case, it would be null (in the sense of the parquet definition level for `foo.typed_value.x.value` not being its maximum value).

Reply:
So I can't have a shredded value that is Nullable?

Reply:
Can't i have foo.typed_value.x.value = {'00'}?

Reply:
I'm not sure I understand the question. That represents a Variant Null, but in terms of Parquet definition levels, it's non-null. So Variant Null (`{"x": null}` in JSON) is represented by the non-null Variant value 0x00 in `x.value`, and a missing value is represented by (definition-level based) null in `x.value` and `x.typed_value`.

Reply:
I asked this lower down, but I should have typed it out here.

Let's say I have a query on Variant column foo
`SELECT * WHERE foo.x = 5`

Now within a file I have variants
1) foo.x : 5
2) foo.x : null (is this possible?)
3) foo.y : "bazz"

Can I shred this file such that I can follow the rules below for file skipping where "foo.typed_value.x.value" is null (real null) for all the above rows

The number 2 case is the one I am worried about, for 3 I obviously just have null for "value" and "typed_value". But For case 2 what do I have?

Reply:
> 2. foo.x : null (is this possible?)
`x` is a field of the Variant value foo, so it can't be null in a SQL/Parquet sense, it can only be Variant Null. That being said, in the Spark implementation (and I think Snowflake is similar) Variant Null will be coerced to a SQL null in most contexts (e.g. cast to string).

So to answer your question, `foo.typed_value.x.value` would be non-null. You'd need to check min/max stats to see if all values in `value` are exactly `0x00` in order to do file skipping.

Before the current PR, it was possible to do this - a missing value was encoded based on the definition level of the field (`x` in this case) and a variant null based on `value` and `typed_value` both being null. @rdblue's view was that it would be better to save a definition level by making the field `x` required (https://github.com/apache/parquet-format/pull/461#discussion_r1811353611), and relying on binary min/max stats to identify this all-variant-null case.

Reply:
@cashmand @rdblue 

following  example from David is not inline with my understanding of this spec :
**_{ "foo" : { }}: both foo.typed_value.x.value and foo.typed_value.x.typed_value for field x are null, indicating that x is missing from the object. But foo.typed_value is non-null, so there is an empty object, not a missing foo.
{ "foo": { "y": 123 }}: Here, x is also missing, so its fields are null, just like the second case. Assuming there's no y in the shredding schema, foo.value would store the binary representation of { "y": 123 }. foo.typed_value should still be non-null to indicate that there is a non-null object, just like in the empty object case._**

For { "foo": { "y": 123 }},  if x is the only shredded field, I do not think it should be necessary to create an empty object for the shredded fields. I guess we can do that, but just to confirm my understanding, this should not be required. Following this logic, for { "foo" : { }}, what is non-null can either be value, typed_value or both as long as all indicate empty objects and hence present values.

Reply:
For the empty object case, I'd argue:
1) both non-null seems valid, if a bit wasteful.
2) Only typed_value non-null should be valid, and is efficient.
3) I'd prefer for only value non-null to be invalid, and don't see any good reason for it to be valid. I gave an example to motivate this in https://github.com/apache/parquet-format/pull/461#discussion_r1830046851: if a query tries to read a field as a struct with a schema that is a subset of the shredding schema (select cast(variant_col:foo as struct<event_type, event_ts>)`, we don't need to fetch `foo`'s `value` column, only `typed_value`. But if typed_value might be null when there's a non-null object, we do need to, or else we'd incorrectly return NULL instead of {NULL, NULL} in that case.

Reply:
What is the semantics of cast(variant_col:foo as struct<event_type, event_ts>) ? Is it strictly requiring the variant to match the struct schema or not? If it requires schema match(and I think it should require a schema match), then assuming no full shredding, in the common case you will  need to read the value field regardless of whether we always write a empty object to typed_value. For example, instead of an object, you might have an integer or string in the value field, and even if we write an empty object to typed_value, you need to read value and figure out if it is something you can really cast to the target struct. So the argument given here only seems to apply to an empty object  which seems to be a corner case.

Reply:
Spark has two cast variants, cast and try_cast. Both consider it valid to cast an object to a struct and drop fields. Whether this is the right choice is a fair question, but let's assume for now that it won't change. I think you're right that for `cast`, we'd need the `value` column to check for errors due to other types. But for `try_cast`, I think we would only need to check the `typed_value` column if we could rely on it setting the definition level based on the value being an object or not.

I don't think it's a huge deal if we need this extra IO, but my preference would be to have clear and limited choices  wherever possible for shredding decisions, so that readers can make optimal choices without risking correctness issues.

Reply:
Sounds good. I would also like to draw your attention to the fact that this issue might also happen for leaf fields. For example you can have a string field field1. A row might have an int and for that row value will be a variant int. If you do foo:field1::number, you need to read the value field and get the int value. Having "some" value for the typed_value would not be useful here. Similarly, for the same shredding scenario, you might have a row with empty object at that field and I think it again makes more sense to put that to the value field.

Reply:
Yes, agreed. The rule (which is meant to be spelled out in this doc, but feel free to suggest clarifications) is:

1) If `typed_value` is a group and the Variant value being shredded is an object, then `typed_value` must be non-null. `value` may also be non-null (specifically, if there are fields that aren't in the shredding schema).
2) In all other cases, at most one of `typed_value` and `value` can be non-null.

Reply:
I think I agree with @cashmand. When shredding an object (`typed_value` is a group), `typed_value` must be non-null. If we can rely on that rule then projections that only require specific fields don't need to read the value.

Reply:
I've also added this:
> Readers can assume that a value is not an object if `typed_value` is null and that `typed_value` field values are correct; that is, readers do not need to read the `value` column if `typed_value` fields satisfy the required fields.